tmpl/wiki/mirage2-to-mirage3.md

```diff
@@ -0,0 +1,1416 @@
+
+## config.ml and generic/default functions
+
+A number of refinements to the configuration language for Mirage were made in the Mirage 2.7.0 release, which incorporated the [functoria](https://github.com/mirage/functoria) DSL.  It was still possible to use the previous calls, however, until Mirage 3.0.0 where support for them was dropped.  Many users whose unikernels predate Mirage 2.7.0 will first notice the functoria improvements in the 3.0.0 release, as they'll finally be forced to use them.
+
+The idiom for deciding which `impl` to use based on the backend is the most noticeable change.  Numerous older `config.ml` files include code like the following, from [mirage-skeleton's `kv_ro` example in commit `21adfc85b124e886d871079f28bd0a868ba3c5fb`](https://github.com/mirage/mirage-skeleton/tree/21adfc85b124e886d871079f28bd0a868ba3c5fb/kv_ro), which attempts to detect which `kv_ro` device to use based on the target chosen with `mirage configure -t` and environment variables:
+
+```
+let mode =
+  let x = try Unix.getenv "FS" with Not_found -> "crunch" in
+  match x with
+  | "fat" -> `Fat
+  | "crunch" -> `Crunch
+  | x -> failwith ("Unknown FS mode: " ^ x )
+
+let fat_ro dir =
+  kv_ro_of_fs (fat_of_files ~dir ())
+
+let disk =
+  match mode, get_mode () with
+  | `Fat   , _     -> fat_ro "t"
+  | `Crunch, `Xen  -> crunch "t"
+  | `Crunch, (`Unix | `MacOSX) -> direct_kv_ro "t"
+```
+
+Functoria improved the situation considerably by including convenience functions which not only have logic for autodetecting which implementations are suitable based on the backend selected, but also other arguments given to `mirage configure`.  The code above can be replaced with:
+
+```
+let disk = generic_kv_ro "t"
+```
+
+`generic_kv_ro` knows about not only the Xen, Unix, and MacOSX targets handled by the previous config.ml code, but also will do the right thing for new targets Ukvm, Virtio, and Qubes.  The command-line argument `--kv_ro` will be understood when `mirage configure` is run from a directory with a `config.ml` including `generic_kv_ro`.
+
+Several useful `default_` and `generic_` functions are provided by [the Mirage module](http://docs.mirage.io/mirage/Mirage) for use in `config.ml`:
+
+```
+~/mirage$ grep -E 'val (generic_|default_)'  lib/mirage.mli
+val default_qubesdb: qubesdb impl
+val default_time: time impl
+val default_posix_clock: pclock impl
+val default_monotonic_clock: mclock impl
+val default_reporter:
+val default_random: random impl
+val default_console: console impl
+val default_io_page: io_page impl
+val generic_kv_ro:
+val default_network: network impl
+val generic_stackv4:
+val default_argv: Functoria_app.argv impl
+```
+
+Many existing unikernels will find it useful to replace their existing network stack detecting code with `generic_stackv4` in particular -- let's see an example.
+
+## porting the stackv4 example to Mirage 3
```

Comment:
oh yes, I noticed that after -- I'd prefer a consistent capitalisation style for the headings in mirage-www content (as it makes the CSS styling easier to look good, among other things)